Remove symengine from the requirements list

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

In #13278 we removed our dependency on symengine and replaced it with our own symbolic engine. However that PR neglected to remove symengine from the requirements list. This commit corrects this oversight and removes it from the requirements list. It also updates the release notes to make them a bit more precise and document the change in requirements and potential behavior differences with the new engine.

### Details and comments

mtreinish committed:
Remove symengine from the requirements list

In #13278 we removed our dependency on symengine and replaced it with
our own symbolic engine. However that PR neglected to remove symengine
from the requirements list. This commit corrects this oversight and
removes it from the requirements list. It also updates the release notes
to make them a bit more prescise and document the change in requirements
and potential behavior differences with the new engine.

mtreinish commented:
I think we can actually drop sympy from the requirements list too. I'll give a try and push up a commit if we can.

EDIT: I looked into this, it is possible to remove but there is some more work needed to address the internals of a couple functions to remove a dependency there. I'll work on this in a parallel PR.

Cryoris commented on qiskit/qpy/binary_io/schedules.py:
```diff
@@ -95,6 +94,7 @@ def _loads_symbolic_expr(expr_bytes, use_symengine=False):
         return common.load_symengine_payload(expr_bytes)
     else:
         from sympy import parse_expr
+        import symengine as sym
```
Do we need to add a check for the optional install of symengine here (or at a higher level)? https://github.com/Qiskit/qiskit/blob/a18e15163d8a93c0e072d4d3c50151a2cd9f036a/qiskit/utils/optionals.py#L333

mtreinish replied:
It should already happen at a higher level since this is pretty deep in the call stack, but I can add the decorator here too just out an abundance of caution it doesn't really hurt.

mtreinish replied:
Actually looking at this again this whole function isn't necessary anymore. Since we don't support pulse all that matters is that we read the appropriate length for old versions that support it. We can't generate a pulse schedule anymore so there is no advantage to actually parsing the data, since we'll just throw it away as not representable in a circuit. I'll just have this always return `None` and we don't need to import it at all.

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/14389/commits/1dcd53817e6b2d6959081747c6bbfadbfda9c454

Cryoris commented on requirements.txt:
```diff
@@ -6,8 +6,3 @@ dill>=0.3
-# If updating the version range here, consider updating the 
-# list of symengine dependencies used in the cross-version tests
-# in 'test/qpy_compat/run_tests.sh' and 'test/qpy_compat/qpy_test_constraints.txt'
-symengine>=0.11,<0.14
```
Since QPY can still rely on `symengine`, do we need to document these constraints somewhere still? E.g. in `requirements-optional` or `constraints`?

mtreinish replied:
`requirements-optional.txt` and constraints is actually only for testing and ci, I can add a new optional variant to the package metadata so users can do `pip install qiskit[qpy-compat]` which is how we would document this. I mostly didn't because qpu parsing will raise an exception with the error message about the symengine version if it detects the mismatch already. But it doesn't hurt to add it there

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14389/commits/dd4a32df746dea033455020b98ca8e95033097eb

mtreinish committed:
Add new optional extra variant for symengine for QPY

mtreinish committed:
Stop parsing sympy and symengine pulse expressions

During QPY deserialization of old payloads that include pulse schedules
we are unable to construct the pulse schedule from the payload. For
circuits that include pulse gates the strategy for loading those
payloads is to emit a warning that the pulse component will be omitted
and create the rest of the circuit like normal. To do this the qpy
deserialization has to read the size of the payload to seek to the right
position. However we don't need to actually parse the expression because
it's never used. This commit removes the parsing step it is
unnecessary after we do the read that will seek the cursor
appropriately.

Cryoris commented on releasenotes/notes/parameter-expression-rs-6fddbb2556b7e1f7.yaml:
```diff
@@ -1,8 +1,29 @@
+    objects has been removed. It has been replaced by a internal symbolic engine and is no
+    longer required for core functionality in Qiskit. The only exception is that symengine
+    was embedded into QPY formats 10, 11, and 12 so it is still required if you are
+    deserializing those formats. The depdency on symengine for :func:`.qpy.load` was made
```
```suggestion
    deserializing those formats. The dependency on symengine for :func:`.qpy.load` was made
```

Cryoris reviewed (commented):
There's a merge conflict, otherwise lgtm!

mtreinish committed:
Update releasenotes/notes/parameter-expression-rs-6fddbb2556b7e1f7.yaml

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mtreinish committed:
Merge remote-tracking branch 'origin/main' into remove-symengine

mtreinish committed:
Remove unused arguments

mtreinish committed:
Merge branch 'main' into remove-symengine

mtreinish committed:
Merge branch 'main' into remove-symengine

Cryoris reviewed (approved):
LGTM, thanks for the updates!
